File: android/app/src/main/java/com/fitlog/app/domain/Body.kt
package com.fitlog.app.domain

object Body {

    enum class Kind(val wire: String) {
        BODY_WEIGHT("body_weight"),
        BODY_FAT("body_fat"),
        WAIST("waist"),
        CHEST("chest"),
        ARM("arm"),
        THIGH("thigh"),
        HIP("hip"),
        NECK("neck"),
        OTHER("other"),
        ;

        companion object {
            fun fromWire(value: String): Kind? = entries.firstOrNull { it.wire == value }
        }
    }

    enum class Unit(val wire: String) {
        KG("kg"),
        CM("cm"),
        PERCENT("%"),
        ;

        companion object {
            fun fromWire(value: String): Unit? = entries.firstOrNull { it.wire == value }
        }
    }

    data class Point(
        val id: String,
        val kind: Kind,
        val measuredAtMs: Long,
        val value: Double,
        val unit: Unit,
        val notes: String?,
    )

    data class Stats(
        val count: Int,
        val first: Double?,
        val latest: Double?,
        val min: Double?,
        val max: Double?,
        val deltaAbs: Double?,
        val deltaPct: Double?,
    )

    fun unitForKind(kind: Kind): Unit = when (kind) {
        Kind.BODY_WEIGHT -> Unit.KG
        Kind.BODY_FAT -> Unit.PERCENT
        else -> Unit.CM
    }

    fun validate(kindWire: String, value: Double): String? {
        val kind = Kind.fromWire(kindWire)
            ?: return "Tipo de medida desconocido: $kindWire"
        if (!value.isFinite() || value <= 0) {
            return "El valor debe ser un número mayor que 0"
        }
        if (kind == Kind.BODY_FAT && value > 100) {
            return "El porcentaje de grasa debe estar entre 0 y 100"
        }
        return null
    }

    fun series(
        metrics: List<Point>,
        kind: Kind,
        range: Progress.Range? = null,
    ): List<Point> = metrics
        .filter { metric ->
            if (metric.kind != kind) return@filter false
            if (range == null) return@filter true
            val from = range.fromMs
            if (from != null && metric.measuredAtMs < from) return@filter false
            metric.measuredAtMs <= range.toMs
        }
        .sortedWith(compareBy({ it.measuredAtMs }, { it.id }))

    fun stats(points: List<Point>): Stats {
        if (points.isEmpty()) {
            return Stats(
                count = 0,
                first = null,
                latest = null,
                min = null,
                max = null,
                deltaAbs = null,
                deltaPct = null,
            )
        }

        val values = points.map { it.value }
        val first = values.first()
        val latest = values.last()
        val deltaAbs = Formulas.roundToTenth(latest - first)

        return Stats(
            count = points.size,
            first = first,
            latest = latest,
            min = values.min(),
            max = values.max(),
            deltaAbs = deltaAbs,
            deltaPct = if (first == 0.0) null else Formulas.roundToTenth(deltaAbs / first * 100),
        )
    }

    fun formatValue(value: Double, unit: Unit): String {
        val rounded = Formulas.roundToTenth(value)
        val text = if (rounded == rounded.toInt().toDouble()) {
            rounded.toInt().toString()
        } else {
            rounded.toString()
        }
        return "$text ${unit.wire}"
    }

    fun label(kind: Kind): String = when (kind) {
        Kind.BODY_WEIGHT -> "Peso corporal"
        Kind.BODY_FAT -> "Grasa corporal"
        Kind.WAIST -> "Cintura"
        Kind.CHEST -> "Pecho"
        Kind.ARM -> "Brazo"
        Kind.THIGH -> "Muslo"
        Kind.HIP -> "Cadera"
        Kind.NECK -> "Cuello"
        Kind.OTHER -> "Otra"
    }
}

File: web/src/domain/body.ts
import { roundToTenth } from '@/domain/formulas';
import type { ProgressRange } from '@/domain/progress';

export const METRIC_KINDS = [
  'body_weight',
  'body_fat',
  'waist',
  'chest',
  'arm',
  'thigh',
  'hip',
  'neck',
  'other',
] as const;

export type MetricKind = (typeof METRIC_KINDS)[number];

export type MetricUnit = 'kg' | 'cm' | '%';

export interface BodyMetricPoint {
  readonly id: string;
  readonly kind: MetricKind;
  readonly measuredAtMs: number;
  readonly value: number;
  readonly unit: MetricUnit;
  readonly notes: string | null;
}

export interface MetricStats {
  readonly count: number;
  readonly first: number | null;
  readonly latest: number | null;
  readonly min: number | null;
  readonly max: number | null;
  readonly deltaAbs: number | null;
  readonly deltaPct: number | null;
}

export function isMetricKind(value: string): value is MetricKind {
  return (METRIC_KINDS as readonly string[]).includes(value);
}

export function unitForKind(kind: MetricKind): MetricUnit {
  if (kind === 'body_weight') return 'kg';
  if (kind === 'body_fat') return '%';
  return 'cm';
}

export function validateMetric(kind: string, value: number): string | null {
  if (!isMetricKind(kind)) {
    return `Tipo de medida desconocido: ${kind}`;
  }
  if (!Number.isFinite(value) || value <= 0) {
    return 'El valor debe ser un número mayor que 0';
  }
  if (kind === 'body_fat' && value > 100) {
    return 'El porcentaje de grasa debe estar entre 0 y 100';
  }
  return null;
}

export function buildMetricSeries(
  metrics: readonly BodyMetricPoint[],
  kind: MetricKind,
  range: ProgressRange | null = null
): BodyMetricPoint[] {
  return metrics
    .filter((metric) => {
      if (metric.kind !== kind) return false;
      if (range === null) return true;
      if (range.fromMs !== null && metric.measuredAtMs < range.fromMs) return false;
      return metric.measuredAtMs <= range.toMs;
    })
    .slice()
    .sort((a, b) => {
      if (a.measuredAtMs !== b.measuredAtMs) return a.measuredAtMs - b.measuredAtMs;
      return a.id < b.id ? -1 : a.id > b.id ? 1 : 0;
    });
}

export function buildMetricStats(points: readonly BodyMetricPoint[]): MetricStats {
  if (points.length === 0) {
    return {
      count: 0,
      first: null,
      latest: null,
      min: null,
      max: null,
      deltaAbs: null,
      deltaPct: null,
    };
  }

  const values = points.map((point) => point.value);
  const first = values[0] as number;
  const latest = values[values.length - 1] as number;
  const deltaAbs = roundToTenth(latest - first);

  return {
    count: points.length,
    first,
    latest,
    min: Math.min(...values),
    max: Math.max(...values),
    deltaAbs,
    deltaPct: first === 0 ? null : roundToTenth((deltaAbs / first) * 100),
  };
}

export function formatMetricValue(value: number, unit: MetricUnit): string {
  const rounded = roundToTenth(value);
  const text = Number.isInteger(rounded) ? String(rounded) : String(rounded);
  return `${text} ${unit}`;
}

export function metricKindLabel(kind: MetricKind): string {
  const labels: Record<MetricKind, string> = {
    body_weight: 'Peso corporal',
    body_fat: 'Grasa corporal',
    waist: 'Cintura',
    chest: 'Pecho',
    arm: 'Brazo',
    thigh: 'Muslo',
    hip: 'Cadera',
    neck: 'Cuello',
    other: 'Otra',
  };
  return labels[kind];
}
